nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -1141,75 +1159,110 @@ impl DataStore {
+                        .select(disk_dsl::id)
+                        .first_async::<Uuid>(conn)
+                        .await
+                        .optional()?;
 
-        if let Some(disk_id) = boot_disk_id {
-            // Ensure the disk is currently attached before updating
-            // the database.
-            let expected_state =
-                api::external::DiskState::Attached(authz_instance.id());
-
-            let attached_disk: Option<Uuid> = disk_dsl::disk
-                .filter(disk_dsl::id.eq(disk_id))
-                .filter(disk_dsl::attach_instance_id.eq(authz_instance.id()))
-                .filter(disk_dsl::disk_state.eq(expected_state.label()))
-                .select(disk_dsl::id)
-                .first_async::<Uuid>(conn)
-                .await
-                .optional()?;
+                    if attached_disk.is_none() {
+                        return Err(err.bail(Error::conflict(
+                            "boot disk must be attached",
+                        )));
+                    }
+                }
 
-            if attached_disk.is_none() {
-                return Err(
-                    err.bail(Error::conflict("boot disk must be attached"))
-                );
+                Ok(())
             }
         }
-        //
-        // NOTE: from this point forward it is OK if we update the
-        // instance's `boot_disk_id` column with the updated value
-        // again. It will have already been assigned with constraint
-        // checking performed above, so updates will just be
-        // repetitive, not harmful.
+    }
+
+    /// Set an instance's CPU count and memory size to the provided values,
+    /// within an existing transaction.
+    async fn instance_set_size_on_conn(
+        &self,
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        err: &OptionalError<Error>,
+        authz_instance: &authz::Instance,
+        ncpus: InstanceCpuCount,
+        memory: ByteCount,
+    ) -> Result<(), diesel::result::Error> {
+        use db::schema::instance::dsl as instance_dsl;
+
+        // Do not allow setting sizes of running instances to ensure that if an
+        // instance is running, its resource usage matches what we record in the
+        // database.
+        const OK_TO_SET_SIZE_STATES: &'static [InstanceState] = &[
+            InstanceState::NoVmm,
+            InstanceState::Failed,
+            InstanceState::Creating,
+        ];
 
         let r = diesel::update(instance_dsl::instance)
             .filter(instance_dsl::id.eq(authz_instance.id()))
-            .filter(instance_dsl::state.eq_any(OK_TO_SET_BOOT_DISK_STATES))
-            .set(instance_dsl::boot_disk_id.eq(boot_disk_id))
+            .filter(instance_dsl::state.eq_any(OK_TO_SET_SIZE_STATES))
+            .filter(
+                instance_dsl::ncpus
+                    .ne(ncpus)
+                    .and(instance_dsl::memory.ne(memory)),
+            )
```

Comment:
hmm, shouldn't this be `or`? doesn't this mean that you can't change _just_ the number of vCPUs or the amount of RAM, you *have* to be changing both?

Reply:
we should probably have tests for changing just vCPUs and just memory, because this seems like a thing we don't want to regress in future...

Reply:
sure should be using `or`, good spot. i've added tests and fixed this in https://github.com/oxidecomputer/omicron/pull/6774/commits/62177576c461a7ad7722b5f5063724caad377b03